### PR TITLE
Added support for streamable embeds

### LINKED ISSUE #49: Support Streamable
By request of @jamessimone: add support for Streamable.

https://streamable.com/documentation

### PR DESCRIPTION
**Pull request by jamessimone**
**What**:
Support for streamable

**Why**:
Closes #49

Streamable is another freely hosted video uploading service

**How**:
Added Streamable.js and tests

**Checklist**:
- [x] Documentation
- [x] Tests
- [x] Ready to be merged

### CODE REVIEW

File: src/Streamable.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+export const shouldTransform = url =>
+  /^https?:\/\/streamable\.com\/s\//.test(url);
+
+export const getHTML = url =>
+  `<div
+  height="0"
+  style="position: relative;"
+  width="100%"
+>
+  <iframe
```
Comment (MichaelDeBoey) [inline src/Streamable.js]: For consistency with other transformers, I'd like to return only the `iframe` and just put it on 1 line here 🙂

File: README.md [hunk 2]
```diff
@@ -184,15 +185,36 @@ https://open.spotify.com/track/0It2bnTdLl2vyymzOkBI3L
...
 ### Twitter
 
 The returned HTML snippet from the Twitter transformer will only be
 automatically recognized as an [Embedded Tweet][embedded-tweet-docs] when
 [Twitter's widget JavaScript][twitter-widget-javascript-docs] is included on the
-page.  
-Since the Twitter transformer doesn't include this JavaScript (because we don't
-want to include it multiple times on a page when having multiple embeds), you
-have to include it yourself. The recommended way of including it is by using
+page. Since the Twitter transformer doesn't include this JavaScript (because we
```
[the remark this answers is not in the document]
Comment (jamessimone) [reply README.md]: Sorry about those changes - prettier did some stuff automatically here and I didn't want to fixup any of my commits once I started making them, this should be resolved now in a series of commits ...

File: src/Streamable.js (hunk 1 above)
Comment (jamessimone) [reply src/Streamable.js]: In the beginning, I was tempted to do just that, but even through the streamable API, but take a look at even what gets returned through the streamable API:

```javascript
{"status": 2, "files": {"mp4": {"status": 2, "width": 1920, "url": "//cdn-b-east.streamable.com/video/mp4/bx960.mp4?token=sNkeIYq-6hxDBOUl20gpTQ&expires=1576186680", "bitrate": 7992605, "duration": 278.511567, "size": 283068434, "framerate": 30, "height": 1080}, "original": {"framerate": 29.97002997002997, "height": 1080, "width": 1920, "duration": 278.511567, "bitrate": 19968979, "size": 700121965}, "mp4-mobile": {"status": 2, "width": 640, "url": "//cdn-b-east.streamable.com/video/mp4-mobile/bx960.mp4?token=grchii_i30Ds1dizRfSt1A&expires=1576186680", "bitrate": 1187597, "duration": 278.5, "size": 43831942, "framerate": 30, "height": 360}}, "embed_code": "<div style=\"width: 100%; height: 0px; position: relative; padding-bottom: 56.25%;\"><iframe class=\"streamable-embed\" src=\"https://streamable.com/o/bx960\" frameborder=\"0\" scrolling=\"no\" style=\"width: 100%; height: 100%; position: absolute;\" allowfullscreen></iframe></div>", "source": null, "thumbnail_url": "//thumbs-east.streamable.com/image/bx960.jpg?height=300", "url": "streamable.com/bx960", "message": null, "title": "Three Good Things", "percent": 100}
```

**Comment by jamessimone** [comment]
@MichaelDeBoey great feedback, thanks for getting back to me!

I don't have any issue retrieving the videos by use of the Streamable API - although their response also includes the wrapper div, please see my comment about that as well. If you're not opposed, I can move forward with making the change to retrieve the embed code by way of their API.

File: src/Streamable.js (hunk 1 above)
Comment (MichaelDeBoey) [reply src/Streamable.js]: Hmmm strange 😕

If you just visit https://api.streamable.com/oembed.json?url=https://streamable.com/moo, you don't see that much info tho...

Comment (jamessimone) [reply src/Streamable.js]: Aha - I see what it is. I was curl'ing `https://api.streamable.com/videos/{shortcode}` - the other endpoint they had provided. I'll use the response from your endpoint!

Comment (MichaelDeBoey) [reply src/Streamable.js]: Very strange those 2 endpoints give different values for the embed code... 😕

**Comment by MichaelDeBoey** [comment]
Some example videos taken from Streamable's homepage:
- https://streamable.com/92ok4
- https://streamable.com/clya3
- https://streamable.com/yktcf

File: src/Streamable.js (hunk 1 above)
Comment (jamessimone) [reply src/Streamable.js]: Haha, I would agree. But such is the way of the world! I just pushed my changes, hopefully they're up to snuff.

File: README.md [hunk 3]
```diff
@@ -20,9 +20,9 @@
 
 Trying to embed well known services (like [CodePen][codepen],
 [CodeSandbox][codesandbox], [Slides][slides], [SoundCloud][soundcloud],
-[Spotify][spotify], [Twitter][twitter] or [YouTube][youtube]) into your
-[Gatsby][gatsby] website can be hard, since you have to know how this needs to
-be done for all of these different services.
+[Spotify][spotify], [Twitter][twitter], [Streamable][streamable] or
```
Comment (MichaelDeBoey) [inline README.md]:
```suggestion
[Spotify][spotify], [Streamable][streamable], [Twitter][twitter] or
```

File: src/Streamable.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+import fetch from 'node-fetch';
+
+export const shouldTransform = url => /^https?:\/\/streamable\.com\//.test(url);
+
+export const getHTML = url =>
+  fetch(`https://api.streamable.com/oembed.json?url=${url}`).then(res =>
+    res.json().then(streamable => streamable.html)
```
Comment (MichaelDeBoey) [inline src/Streamable.js]: Just return `res.json` and add the `then` on the `fetch`

```js
fetch(`https://api.streamable.com/oembed.json?url=${url}`)
  .then(res => res.json())
  .then(({ html }) => html);
```

Even tho we don't need to do a fetch call
We can just return the `iframe` ourselves and save us a fetch call
```html
<iframe src="https://streamable.com/o/moo" frameborder="0" scrolling="no" width="100%" height="480" allowfullscreen></iframe>
```

Comment (MichaelDeBoey) [reply src/Streamable.js]: What your opinion on this one (fetch vs. doing it ourselves) @kentcdodds? 🤔

File: src/__tests__/Streamable.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    },
+    'streamable-url-with-valid-embed-ending': {
+      url: 'https://streamable.com/bx960',
+      valid: true,
+    },
+  }
+);
+
+test('Builds the correct Streamable wrapper and iframe', async () => {
+  const html = await getHTML('https://streamable.com/bx960');
```
Comment (MichaelDeBoey) [inline src/__tests__/Streamable.js]: If we keep doing the `fetch` call, we need to mock it in our tests

File: src/__tests__/__fixtures__/Streamable.md [hunk 6]
```diff
@@ -0,0 +1 @@
+https://streamable.com/s/bx960/hcobuo
```
Comment (MichaelDeBoey) [inline src/__tests__/__fixtures__/Streamable.md]: Kitchensink test should be added too in order to test this one.

File: src/Streamable.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+import fetch from 'node-fetch';
+
+export const shouldTransform = url => /^https?:\/\/streamable\.com\//.test(url);
```
Comment (MichaelDeBoey) [inline src/Streamable.js]: We now have had the following possible links:
- `https://streamable.com/{shortlink}`
- `https://streamable.com/o/{shortlink}` (see `iframe` returned from `oembed.json` call)
- `https://streamable.com/s/{shortlink}`

So I think we should support all of these

If you look at the 3 video's from their homepage (see https://github.com/MichaelDeBoey/gatsby-remark-embedder/pull/48#issuecomment-565169351) and do a right-click to get the embed-code, you also get links that look like:
- `https://streamable.com/s/92ok4/bxmati`
  Is the video same as
  - `https://streamable.com/s/92ok4`
  - `https://streamable.com/92ok4`
- `https://streamable.com/s/clya3/vguntp`
  Is the video same as
  - `https://streamable.com/s/clya3`
  - `https://streamable.com/clya3`
- `https://streamable.com/s/yktcf/jjveaw`
  Is the video same as
  - `https://streamable.com/s/yktcf`
  - `https://streamable.com/yktcf`

File: src/Streamable.js (hunk 4 above)
Comment (kentcdodds) [reply src/Streamable.js]: If they ever change the code that's required to embed then doing the fetch call will be a good thing, so I say let's do that.

File: src/__tests__/Streamable.js (hunk 5 above)
Comment (jamessimone) [reply src/__tests__/Streamable.js]: Sure - if the API changes, though, the test would fail properly. Mocking it in the response will mask this. Your thoughts? I should mention that I'm a huge proponent of using mocking to prevent API calls; I didn't mock in this case because I saw there were other tests also making API calls.

Comment (MichaelDeBoey) [reply src/__tests__/Streamable.js]: Twitter transformer tests are being mocked.
https://github.com/MichaelDeBoey/gatsby-remark-embedder/blob/8fd0afc46da27fb9d6d8dedb5541bf6e8785f686/src/__tests__/Twitter.js#L6-L15
https://github.com/MichaelDeBoey/gatsby-remark-embedder/blob/8fd0afc46da27fb9d6d8dedb5541bf6e8785f686/src/__tests__/plugin.js#L90-L97

**Comment by jamessimone** [comment]
For sure. Much more a fan of rebase, I’ll get on that tomorrow probably. I’ll update you afterwards. Thanks!

File: src/__tests__/transformers/Streamable.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+      valid: true,
+    },
+    'streamable-url-with-long-embed-ending': {
+      url: 'https://streamable.com/s/bx960/hcobuo',
+      valid: true,
+    },
+  }
+);
+
+test('Builds the correct Streamable wrapper and iframe', async () => {
```
Comment (MichaelDeBoey) [inline src/__tests__/transformers/Streamable.js]:
```suggestion
test('Gets the correct Streamable iframe', async () => {
```

File: src/__tests__/plugin.js [hunk 9]
```diff
@@ -13,8 +13,13 @@ describe('gatsby-remark-embedder', () => {
   });
 
   test('can transform all supported links (kitchensink)', async () => {
+    const streamableEmbed = `<iframe class="streamable-embed" src="https://streamable.com/o/bx960" frameborder="0" scrolling="no" width="1920" height="1080" allowfullscreen></iframe>`;
     mockCache({
       'https://twitter.com/kentcdodds/status/1078755736455278592': `<blockquote class="twitter-tweet-from-cache"><p lang="en" dir="ltr">example</p>&mdash; Kent C. Dodds (@kentcdodds) <a href="https://twitter.com/kentcdodds/status/1078755736455278592?ref_src=twsrc%5Etfw">December 28, 2018</a></blockquote>`,
+      'https://streamable.com/s/bx960/hcobuo': streamableEmbed,
```
Comment (MichaelDeBoey) [inline src/__tests__/plugin.js]: Only keep 1 cache  for the kitchensink test & inline `streamableEmbed`

File: src/__tests__/transformers/Streamable.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
+import cases from 'jest-in-case';
+
+import { getHTML, shouldTransform } from '../../transformers/Streamable';
+
+jest.mock('node-fetch', () =>
```
Comment (MichaelDeBoey) [inline src/__tests__/transformers/Streamable.js]: Check the Twitter tests to make this a bit cleaner (can be extracted in another PR)

File: src/__tests__/__fixtures__/Streamable.md (hunk 6 above)
Comment (MichaelDeBoey) [inline src/__tests__/__fixtures__/Streamable.md]: Add all supported variants in the fixture

File: src/__tests__/transformers/Streamable.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  }
+);
+
+test('Gets the correct Streamable iframe', async () => {
+  const html = await getHTML('https://streamable.com/bx960');
+
+  expect(html).toEqual(
+    `<iframe class="streamable-embed" src="https://streamable.com/o/bx960" frameborder="0" scrolling="no" width="1920" height="1080" allowfullscreen></iframe>`
+  );
+});
```
Comment (MichaelDeBoey) [inline src/__tests__/transformers/Streamable.js]: Don't forget to add the plugin test 🙂

File: src/__tests__/helpers/mockFetch.js [hunk 12]
```diff
@@ -0,0 +1,9 @@
+import fetchMock from 'node-fetch';
+
+export const mockFetch = html => {
```
Comment (MichaelDeBoey) [inline src/__tests__/helpers/mockFetch.js]: Please don't extract `mockFetch` in this PR.
We'll do this in another PR, to keep this PR as clean as possible. 🙂

File: src/__tests__/transformers/Streamable.js [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    },
+    'streamable-url-with-long-embed-ending': {
+      url: 'https://streamable.com/s/bx960/hcobuo',
+      valid: true,
+    },
+  }
+);
+
+test('Gets the correct Streamable iframe', async () => {
+  mockFetch(expectedHtmlResult);
```
Comment (MichaelDeBoey) [inline src/__tests__/transformers/Streamable.js]: Inline `expectedHtmlResult` with a separate `class` for the `iframe` & `plugin` test, to be sure we're mocking everything correctly (see Twitter test).

File: src/__tests__/transformers/Streamable.js [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    'streamable-url-with-long-embed-ending': {
+      url: 'https://streamable.com/s/bx960/hcobuo',
+      valid: true,
+    },
+  }
+);
+
+test('Gets the correct Streamable iframe', async () => {
+  mockFetch(
+    `<iframe class="mocked-response" src="https://streamable.com/o/bx960" frameborder="0" scrolling="no" width="1920" height="1080" allowfullscreen></iframe>`
```
Comment (MichaelDeBoey) [inline src/__tests__/transformers/Streamable.js]:
```suggestion
    `<iframe class="streamable-embed-mocked-fetch-transformer" src="https://streamable.com/o/bx960" frameborder="0" scrolling="no" width="1920" height="1080" allowfullscreen></iframe>`
```

File: src/__tests__/transformers/Streamable.js [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+  const html = await getHTML('https://streamable.com/bx960');
+
+  expect(html).toMatchInlineSnapshot(
+    `"<iframe class=\\"mocked-response\\" src=\\"https://streamable.com/o/bx960\\" frameborder=\\"0\\" scrolling=\\"no\\" width=\\"1920\\" height=\\"1080\\" allowfullscreen></iframe>"`
+  );
+});
+
+test('Plugin correctly transforms Streamable links', async () => {
+  mockFetch(
+    `<iframe class="mocked-response" src="https://streamable.com/o/bx960" frameborder="0" scrolling="no" width="1920" height="1080" allowfullscreen></iframe>`
```
Comment (MichaelDeBoey) [inline src/__tests__/transformers/Streamable.js]:
```suggestion
    `<iframe class="streamable-embed-mocked-fetch-plugin" src="https://streamable.com/o/bx960" frameborder="0" scrolling="no" width="1920" height="1080" allowfullscreen></iframe>`
```

File: src/transformers/Streamable.js [hunk 16]
```diff
@@ -0,0 +1,9 @@
+import fetch from 'node-fetch';
+
+export const shouldTransform = url =>
+  /^https?:\/\/streamable\.com\/?[|s|o|\D].+/.test(url);
+
+export const getHTML = url =>
+  fetch(`https://api.streamable.com/oembed.json?url=${url}`)
+    .then(res => res.json())
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]:
```suggestion
    .then(({ json }) => json())
```

File: src/transformers/Streamable.js [hunk 17]
```diff
@@ -0,0 +1,9 @@
+import fetch from 'node-fetch';
+
+export const shouldTransform = url =>
+  /^https?:\/\/streamable\.com\/?[|s|o|\D].+/.test(url);
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: Make this more readable by using the `url` dependency, just like we do in all our other transformers:
- Check if the `host` is 'streamable.com' or 'www.streamable.com'
- Check the length of the `/` splitted `pathname`
  - If `1`
    - Return `true`
  - If `2` or `3`
    - Return if the first part is one of `'/o/'` or `'/s/'` (copy over the `includesSomeOfArray` from the Spotify transformer, we'll extract that into a helper function in a separate PR)
  - Else
    - Return `false`

File: README.md [hunk 18]
```diff
@@ -184,6 +185,36 @@ https://open.spotify.com/track/0It2bnTdLl2vyymzOkBI3L
...
 
+### Streamable
+
+#### Usage
+
+```md
+https://streamable.com/yourShortCodeHere
+```
+
+Or you can use one of the three other valid Streamable link formats:
```
Comment (MichaelDeBoey) [inline README.md]: Remove this part, as we only include one example in the docs

File: README.md [hunk 19]
```diff
@@ -184,6 +185,36 @@ https://open.spotify.com/track/0It2bnTdLl2vyymzOkBI3L
 ></iframe>
 ```
 
+### Streamable
+
+#### Usage
+
+```md
+https://streamable.com/yourShortCodeHere
```
Comment (MichaelDeBoey) [inline README.md]:
```suggestion
https://streamable.com/s/bx960
```

File: README.md [hunk 20]
```diff
@@ -184,6 +185,36 @@ https://open.spotify.com/track/0It2bnTdLl2vyymzOkBI3L
...
+https://streamable.com/o/yourShortCodeHere
+https://streamable.com/s/yourShortCodeHere/anotherUrlSectionStreamableSometimesIncludes
+```
+
+#### Result
+
+```html
+<iframe
+  class="streamable-embed"
+  src="https://streamable.com/s/some/embed-code-video-here"
```
Comment (MichaelDeBoey) [inline README.md]:
```suggestion
  src="https://streamable.com/o/bx960"
```

File: README.md [hunk 21]
```diff
@@ -184,6 +185,36 @@ https://open.spotify.com/track/0It2bnTdLl2vyymzOkBI3L
...
+  frameborder="0"
+  scrolling="no"
+  width="your width here%"
```
Comment (MichaelDeBoey) [inline README.md]:
```suggestion
  width="1920"
```

File: README.md [hunk 22]
```diff
@@ -184,6 +185,36 @@ https://open.spotify.com/track/0It2bnTdLl2vyymzOkBI3L
...
+  height="your height here"
```
Comment (MichaelDeBoey) [inline README.md]:
```suggestion
  height="1080"
```

**Comment by jamessimone** [comment]
@MichaelDeBoey pushed some more changes up.

File: src/transformers/Streamable.js [hunk 23]
```diff
@@ -0,0 +1,27 @@
+import { URL } from 'url';
+
+import fetch from 'node-fetch';
+
+const includesSomeOfArray = (string, array) =>
+  array.some(item => string.includes(item));
+
+export const shouldTransform = url => {
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: This function can be simplified/made a bit less messy (imo) to

```js
export const shouldTransform = url => {
  const { host, pathname } = new URL(url);
  const trimmedPathName = getTrimmedPathName(pathname);

  return ['streamable.com', 'www.streamable.com'].includes(host) &&
    trimmedPathName.length > 0 &&
    trimmedPathName.length < 3 &&
    (trimmedPathName.length === 1 ||
      includesSomeOfArray(pathname, ['/o/', '/s/']));
};
```

You could just copy over the `getTrimmedPathName` function from the Slides transformer for now

**Comment by jamessimone** [comment]
@MichaelDeBoey yes I agree that the function could be simplified. I would not write something like this - I am trying to implement things exactly as you are describing. There were two issues with the function that you pasted in your last comment which I have resolved, to the best of my ability. I don't write functions with deeply nested return statements like this (particularly with parenthetical or statements), so it's possible that you'll want further changes made.

File: src/transformers/Streamable.js [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+  const { host, pathname } = new URL(url);
+  const trimmedPathName = getTrimmedPathName(pathname).split('/');
+
+  return (
+    ['streamable.com', 'www.streamable.com'].includes(host) &&
+    !includesSomeOfArray(pathname, [
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: Only check if `pathname` is included in array when `trimmedPathName.length === 1`

File: src/transformers/Streamable.js [hunk 25]
```diff
@@ -0,0 +1,32 @@
...
+      'documentation',
+      'login',
+      'recover',
+      'settings',
+      'signup',
+    ]) &&
+    trimmedPathName.length > 0 &&
+    trimmedPathName.length <= 3 &&
+    (trimmedPathName.length === 1 ||
+      includesSomeOfArray(pathname, ['/e/', '/g/', '/o/', '/s/', '/t/']))
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: Check if `trimmedPathName[0]` is included in array instead so it's more strict what we're checking

File: src/transformers/Streamable.js [hunk 26]
```diff
@@ -11,18 +11,14 @@ export const shouldTransform = url => {
...
-    (trimmedPathName.length === 1 ||
-      includesSomeOfArray(pathname, ['/e/', '/g/', '/o/', '/s/', '/t/']))
+    (['streamable.com', 'www.streamable.com'].includes(host) &&
+      trimmedPathName.length === 1 &&
+        !['documentation', 'login', 'recover', 'settings', 'signup'].includes(
+          trimmedPathName[0]
+        )) ||
+    (trimmedPathName.length > 1 &&
+      trimmedPathName.length <= 3 &&
+      includesSomeOfArray(trimmedPathName[0], ['e', 'g', 'o', 's', 't']))
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: This can just be

```js
array.includes(trimmedPathName[0])
```

File: src/transformers/Streamable.js [hunk 27]
```diff
@@ -11,18 +11,14 @@ export const shouldTransform = url => {
...
-      'settings',
-      'signup',
-    ]) &&
-    trimmedPathName.length > 0 &&
-    trimmedPathName.length <= 3 &&
-    (trimmedPathName.length === 1 ||
-      includesSomeOfArray(pathname, ['/e/', '/g/', '/o/', '/s/', '/t/']))
+    (['streamable.com', 'www.streamable.com'].includes(host) &&
+      trimmedPathName.length === 1 &&
+        !['documentation', 'login', 'recover', 'settings', 'signup'].includes(
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: This can just be

```js
array.includes(pathname)
```

File: src/transformers/Streamable.js [hunk 28]
```diff
@@ -0,0 +1,25 @@
+import { URL } from 'url';
+import fetch from 'node-fetch';
+
+const getTrimmedPathName = pathname => pathname.replace(/^\/|\/+$/g, '');
+
+export const shouldTransform = url => {
+  const { host, pathname } = new URL(url);
+  const trimmedPathName = getTrimmedPathName(pathname).split('/');
+
+  return (
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: I think you should return the following to express more the meaning of what we're checking:

```js
  return (
    ['streamable.com', 'www.streamable.com'].includes(host) &&
    trimmedPathName.length > 0 &&
    trimmedPathName.length <= 3 &&
    (
      (
        trimmedPathName.length === 1 &&
        !['/documentation', '/login', '/recover', '/settings', '/signup'].includes(pathname)
      ) ||
      ['e', 'g', 'o', 's', 't'].includes(trimmedPathName[0])
    )
  );
```

**Comment by MichaelDeBoey** [comment]
@jamessimone I just tried out the following original links and they all work, but adding them to the embed url, they all give a `No video found` error back.

|original|embed|
|--|--|
|https://streamable.com/e/moo|https://api.streamable.com/oembed.json?url=https://streamable.com/e/moo|
|https://streamable.com/e/moo/username|https://api.streamable.com/oembed.json?url=https://streamable.com/e/moo/usernam|
|https://streamable.com/g/moo|https://api.streamable.com/oembed.json?url=https://streamable.com/g/moo|
|https://streamable.com/g/moo/username|https://api.streamable.com/oembed.json?url=https://streamable.com/g/moo/username|
|https://streamable.com/o/moo|https://api.streamable.com/oembed.json?url=https://streamable.com/o/moo|
|https://streamable.com/o/moo/username|https://api.streamable.com/oembed.json?url=https://streamable.com/o/moo/username|
|https://streamable.com/s/moo|https://api.streamable.com/oembed.json?url=https://streamable.com/s/moo|
|https://streamable.com/s/moo/username|https://api.streamable.com/oembed.json?url=https://streamable.com/s/moo/username|
|https://streamable.com/t/moo|https://api.streamable.com/oembed.json?url=https://streamable.com/t/moo|
|https://streamable.com/t/moo/username|https://api.streamable.com/oembed.json?url=https://streamable.com/t/moo/username|

I'd like to support all of them tho, so my suggestion would be to add a `getNormalizedStreamableUrl` function, which normalizes the links to be `https://streamable.com/{videoID}`.
Don't forget to add tests for them 😉

File: src/transformers/Streamable.js [hunk 29]
```diff
@@ -1,6 +1,13 @@
 import { URL } from 'url';
 import fetch from 'node-fetch';
 
+export const getNormalizedStreamableUrl = url => {
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: I think doing it like this, will make the meaning of the code less descriptive and more declaring what we're doing in the code.

```js
export const getNormalizedStreamableUrl = url => {
  const { pathname } = new URL(url);
  const trimmedPathName = getTrimmedPathName(pathname).split('/');

  if (trimmedPathName.length === 1) {
    return url;
  }

  return `https://streamable.com/${trimmedPathName[1]}`;
};
```

File: src/transformers/Streamable.js [hunk 30]
```diff
@@ -1,6 +1,13 @@
...
+  const splitUrlSegments = url.split('/');
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: I think doing it like this, will make the meaning of the code less descriptive and more declaring what we're doing in the code.

```js
export const getNormalizedStreamableUrl = url => {
  const { pathname } = new URL(url);
  const trimmedPathName = getTrimmedPathName(pathname).split('/');

  if (trimmedPathName.length === 1) {
    return url;
  }

  return `https://streamable.com/${trimmedPathName[1]}`;
};
```

File: src/transformers/Streamable.js [hunk 31]
```diff
@@ -24,6 +31,10 @@ export const shouldTransform = url => {
 };
 
 export const getHTML = url =>
-  fetch(`https://api.streamable.com/oembed.json?url=${url}`)
+  fetch(
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: I think that we should extract the `normalizedUrl` var here

File: src/__tests__/transformers/Streamable.js [hunk 32]
```diff
@@ -0,0 +1,243 @@
...
+  fetchMock.mockResolvedValue({ json: () => Promise.resolve({ html }) });
+
+beforeEach(() => {
+  fetchMock.mockClear();
+});
+
+cases(
+  'normalized streamable path validation',
+  ({ url, shortCode }) => {
+    expect(getNormalizedStreamableUrl(url)).toBe(shortCode);
```
Comment (MichaelDeBoey) [inline src/__tests__/transformers/Streamable.js]: We should test the normalized url here, not the `shortCode`, because otherwise we're doing a call to `https://api.streamable.com/oembed.json?url=moo` and that's not valid

File: src/transformers/Streamable.js [hunk 33]
```diff
@@ -0,0 +1,40 @@
...
+export const shouldTransform = url => {
+  const { host, pathname } = new URL(url);
+  const trimmedPathName = getTrimmedPathName(pathname).split('/');
+
+  return (
+    ['streamable.com', 'www.streamable.com'].includes(host) &&
+    trimmedPathName.length > 0 &&
+    trimmedPathName.length <= 3 &&
+    ((trimmedPathName.length === 1 &&
+      ![
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: Maybe also extract the array to a `ignoredPaths` variable, like we have `podcastTypes` in the Spotify transformer

File: src/transformers/Streamable.js [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+        '/documentation',
+        '/login',
+        '/recover',
+        '/settings',
+        '/signup',
+      ].includes(pathname)) ||
+      ['e', 'g', 'o', 's', 't'].includes(trimmedPathName[0]))
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: Maybe also extract this array to a `possibleCachePaths` variable so it less of a magic array? 🤔

File: src/transformers/Streamable.js [hunk 35]
```diff
@@ -0,0 +1,44 @@
...
+
+const getTrimmedPathName = pathname => pathname.replace(/^\/|\/+$/g, '');
+
+export const getNormalizedStreamableUrl = url => {
+  const { pathname } = new URL(url);
+  const streamableEmbedUrl = 'https://api.streamable.com/oembed.json?url=';
+  const trimmedPathName = getTrimmedPathName(pathname).split('/');
+
+  if (trimmedPathName.length === 1) {
+    return `${streamableEmbedUrl}${url}`;
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]: The Streamable embed url shouldn’t be attached

```suggestion
    return url;
```

File: src/transformers/Streamable.js [hunk 36]
```diff
@@ -0,0 +1,44 @@
...
+  }
+
+  return `${streamableEmbedUrl}https://streamable.com/${trimmedPathName[1]}`;
```
Comment (MichaelDeBoey) [inline src/transformers/Streamable.js]:
```suggestion
  return `https://streamable.com/${trimmedPathName[1]}`;
```

File: src/transformers/Streamable.js (hunk 31 above)
Comment (MichaelDeBoey) [reply src/transformers/Streamable.js]:
```js
const normalizedUrl = getNormalizedUrl(url);

return fetch(`https://api.streamable.com/oembed.json?url=${normalizedUrl}`)
```

**Comment by MichaelDeBoey** [comment]
@all-contributors Please add @jamessimone for code and tests
